### PR TITLE
test: add retries to dlp tests

### LINKED ISSUE #7728: dlp.snippets.InspectTests: testInspectStringWithExclusionDictSubstring failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 24e97c8966c3f2af5c16fc1630bff70db664e752
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/789c3d20-dad7-4ee3-960d-8d9d4b193109), [Sponge](http://sponge2/789c3d20-dad7-4ee3-960d-8d9d4b193109)
status: failed
<details><summary>Test output</summary><br><pre>com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:110)
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:41)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:86)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:67)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1132)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1270)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1038)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:808)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:574)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:544)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:535)
	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:471)
	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:435)
	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:468)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:563)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:70)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:744)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:723)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.dlp.v2.DlpServiceClient.inspectContent(DlpServiceClient.java:285)
		at dlp.snippets.InspectStringWithExclusionDictSubstring.inspectStringWithExclusionDictSubstring(InspectStringWithExclusionDictSubstring.java:110)
		at dlp.snippets.InspectTests.testInspectStringWithExclusionDictSubstring(InspectTests.java:143)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
		at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
		at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
		at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
		at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
		at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
		at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
		at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
		at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
		at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
		at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
		at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
		at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
		at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
		at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
		at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
		at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
		at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
		at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
		at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
		at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
		at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
		at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
		at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
		at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
		at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
		at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
		at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
	at io.grpc.Status.asRuntimeException(Status.java:539)
	... 17 more
</pre></details>

### PR DESCRIPTION
**Pull request by kweinmeister**
Fixes #7728

### CODE REVIEW

File: dlp/snippets/src/test/java/dlp/snippets/TestBase.java [hunk 1]
```diff
@@ -18,14 +18,17 @@
...
 import com.google.common.collect.ImmutableList;
 import java.io.ByteArrayOutputStream;
 import java.io.PrintStream;
 import org.junit.After;
 import org.junit.Before;
+import org.junit.Rule;
 
 /** Common base class for DLP snippet tests */
 abstract class TestBase {
+  @Rule public final MultipleAttemptsRule multipleAttemptsRule = new MultipleAttemptsRule(5);
```
Comment (realjordanna) [inline dlp/snippets/src/test/java/dlp/snippets/TestBase.java:31]: add a comment as to why this is needed and why 5 was chosen. it's okay if your answer is 5 is a guess, or there is some known upper limit on what it should be...

**Review by realjordanna — CHANGES_REQUESTED** [review]

**Comment by kweinmeister** [comment]
> This sample was flaky due to the `dlp.inspectContent(request)` asynchronous task failed. It may be better to show users how to handle that error than just rerun the test.

That works for me. I can remove the "Fixes" link, if we don't want to resolve the issue with this improvement.

@realjordanna any thoughts? Would your team be able to also address the more granular issue as @averikitsch described?

**Comment by realjordanna** [comment]
That makes sense to me ... I'm looking through the flake at https://source.cloud.google.com/results/invocations/789c3d20-dad7-4ee3-960d-8d9d4b193109/targets/cloud-devrel%2Fjava%2Fjava-docs-samples%2Fjava8%2Fperiodic/log, and the error in the failing test isn't useful because it is truncated. I can't actually see what the server returned to the test because of the truncation. Catching the exception would allow us to get a better log statement to see what's going on. If we have the project_number from the test I could grab logs internally possibly, but this method is not known to be flaky on the backend side. I have identical http based tests that look like this that are not flaking. Do we know what the deadline is set too? Maybe it's too short?

**Comment by kweinmeister** [comment]
Thank you @realjordanna and @averikitsch. We can merge this for now, and if see we continued flakiness, we can do a more thorough investigation.

**Pull request merged**